lib/SILOptimizer/Mandatory/TFConstExpr.cpp

```diff
@@ -1557,3 +1642,133 @@ computeConstantValues(ArrayRef<SILValue> values,
     numInstEvaluated = 0;
   }
 }
+
+/// Analyze the array users of an _allocateUninitialized call, to see if they
+/// are all simple things we can remove.  If so, add them all to arrayInsts and
+/// return false.  If not, return true.
+static bool analyzeArrayInitUses(SILValue v,
+                                 SmallPtrSet<SILInstruction*, 8> *arrayInsts) {
+  for (auto *use : v->getUses()) {
+    auto *user = use->getUser();
+
+    // We can always remove retain/release instructions and debug_value.
+    if (isa<StrongRetainInst>(user) || isa<StrongReleaseInst>(user) ||
+        isa<DebugValueInst>(user)) {
+      if (arrayInsts) arrayInsts->insert(user);
+      continue;
+    }
+
+    // We can look through unpacking and repacking of structs that eventually
+    // turn into a retain or release.
+    if (isa<StructExtractInst>(user) || isa<StructInst>(user)) {
+      if (arrayInsts) arrayInsts->insert(user);
+      if (analyzeArrayInitUses(user->getResults()[0], arrayInsts))
+        return true;
+      continue;
+    }
+
+    // Oops we found an unknown use!
+    return true;
+  }
+  return false;
+}
+
+/// Try to decode the specified apply of the _allocateUninitializedArray
+/// function in the standard library.  This attempts to figure out what the
+/// resulting elements will be.  This fills in the elements result and returns
+/// true on success.
```

Comment:
some other existing functions return "true" on failure. 

My first preference is to "return true on success" consistently throughout the code base, but i understand that can involve changing large parts of the code base and may not be worthwhile.

My fallback preference is to "return true on failure" for consistency reasons.

What do you think?

Reply:
+1

Reply:
Sure, will change.